# Marketing Model Mix
This file is just a sample working on marketing mix model. It is based on hypothetical data. The emphasis here is on methodology and understanding procedure rather than output

Installing and importing few important libraries

In [5]:
    pip install plotly 

^C
Note: you may need to restart the kernel to use updated packages.


In [98]:
import pandas as pd
import plotly.express as px
import statsmodels as sm
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

Importing the data

In [157]:
dataframe = pd.read_csv('dataset_real.csv')

In [145]:
dataframe.info

<bound method DataFrame.info of     ordine_data  revenue  google_search_spent  google_performance_max_spent  \
0    12-09-2021  2461.81               151.48                         88.20   
1    13-09-2021  6918.33               261.70                         90.56   
2    14-09-2021  5101.67               207.53                         74.14   
3    15-09-2021  5335.02               204.25                         65.01   
4    16-09-2021  4384.38               189.94                         69.26   
..          ...      ...                  ...                           ...   
158  17-02-2022  6061.37               347.75                        125.32   
159  18-02-2022  3722.15               258.10                         79.30   
160  19-02-2022  2670.47                92.04                          9.96   
161  20-02-2022  3385.87                86.37                         40.50   
162  21-02-2022   372.27                42.17                         22.04   

     fb_retargeting

Manuplating the data

In [102]:
x= []
for y in range (163):
    x.append('week' + str(y))

In [103]:
dataframe["week"]=x

In [104]:
dataframe.drop('ordine_data', inplace=True, axis=1)

In [105]:
dataframe.head()

revenue  google_search_spent  google_performance_max_spent  \
0  2461.81               151.48                         88.20   
1  6918.33               261.70                         90.56   
2  5101.67               207.53                         74.14   
3  5335.02               204.25                         65.01   
4  4384.38               189.94                         69.26   

   fb_retargeting_spent  fb_prospecting_spent  google_organico   week  
0                 12.15                 31.48             2603  week0  
1                 13.62                 29.61             5691  week1  
2                 12.69                 30.09             5616  week2  
3                 11.95                 30.08             5676  week3  
4                 13.54                 31.03             5398  week4

Understanding the data on graph

In [106]:
fig = px.line(dataframe, x='week', y = dataframe.columns)
fig.show()

We can already see that expenditure on 'google_organico' is highly co-orelated with the revenue

In [107]:
dataframe.corr()

revenue  google_search_spent  \
revenue                       1.000000             0.448790   
google_search_spent           0.448790             1.000000   
google_performance_max_spent  0.198876             0.597335   
fb_retargeting_spent          0.215785             0.334000   
fb_prospecting_spent          0.183129             0.247573   
google_organico               0.697560             0.552363   

                              google_performance_max_spent  \
revenue                                           0.198876   
google_search_spent                               0.597335   
google_performance_max_spent                      1.000000   
fb_retargeting_spent                              0.196233   
fb_prospecting_spent                              0.304486   
google_organico                                   0.234278   

                              fb_retargeting_spent  fb_prospecting_spent  \
revenue                                   0.215785              0.183129   
google_search_spent                       0.334000              0.247573   
google_performance_max_spent              0.196233              0.304486   
fb_retargeting_spent                      1.000000              0.855595   
fb_prospecting_spent                      0.855595              1.000000   
google_organico                           0.102636              0.040979   

                              google_organico  
revenue                              0.697560  
google_search_spent                  0.552363  
google_performance_max_spent         0.234278  
fb_retargeting_spent                 0.102636  
fb_prospecting_spent                 0.040979  
google_organico                      1.000000

In [108]:
fig1 = px.scatter(dataframe, x='fb_prospecting_spent', y='revenue') 
fig2 = px.scatter(dataframe, x='fb_retargeting_spent', y='revenue')
fig3 = px.scatter(dataframe, x='google_performance_max_spent', y='revenue')
fig4 = px.scatter(dataframe, x='google_search_spent', y='revenue')
fig5 = px.scatter(dataframe, x='google_organico', y='revenue')
fig6 = px.scatter(dataframe, x='google_performance_max_spent', y='revenue')

In [109]:
fig1.show()
fig2.show()
fig3.show()
fig4.show()
fig5.show()
fig6.show()

We can observe that fb_prospecting_spent, fb_retargeting_spent, google_performance_max_spent, and google_search_spent are almost horizontal line. They show lesser positive correlation with revenue 

In [110]:
variables = ['google_search_spent','google_performance_max_spent','fb_retargeting_spent','fb_prospecting_spent','google_organico']

In [111]:
import statsmodels.api as sm

In [112]:
X=dataframe[variables]
y=dataframe['revenue']
X = sm.add_constant(X)

result = sm.OLS(y,X).fit()

print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                revenue   R-squared:                       0.513
Model:                            OLS   Adj. R-squared:                  0.498
Method:                 Least Squares   F-statistic:                     33.09
Date:                Tue, 20 Sep 2022   Prob (F-statistic):           5.82e-23
Time:                        21:49:07   Log-Likelihood:                -1483.7
No. Observations:                 163   AIC:                             2979.
Df Residuals:                     157   BIC:                             2998.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                                   coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------
const           

C:\Users\Ankit\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning:

In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only



Here Adjusted R-squared is .513; which means that only ~50% of variation in the revenue can be explained by the marketing expenditure data that we have provided. Rest of the vairaition is still unexplained.

If we look at the statistical significance of the independent variables then we observe that all except google_organico are of low statistical significance.

Hence we now drop indepedent variables with lowest statistical significance first in next model

In [113]:
variables = ['google_search_spent','google_organico','google_performance_max_spent','fb_prospecting_spent']
X=dataframe[variables]
y=dataframe['revenue']
X = sm.add_constant(X)

result = sm.OLS(y,X).fit()

print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                revenue   R-squared:                       0.513
Model:                            OLS   Adj. R-squared:                  0.501
Method:                 Least Squares   F-statistic:                     41.62
Date:                Tue, 20 Sep 2022   Prob (F-statistic):           8.46e-24
Time:                        21:50:32   Log-Likelihood:                -1483.7
No. Observations:                 163   AIC:                             2977.
Df Residuals:                     158   BIC:                             2993.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                                   coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------
const           

C:\Users\Ankit\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning:

In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only



google_performance_max_spent and google_search_spent  are still statistically insignificant based on their P-values

In [114]:
variables = ['google_organico','fb_prospecting_spent']
X=dataframe[variables]
y=dataframe['revenue']
X = sm.add_constant(X)

result = sm.OLS(y,X).fit()

print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                revenue   R-squared:                       0.511
Model:                            OLS   Adj. R-squared:                  0.504
Method:                 Least Squares   F-statistic:                     83.44
Date:                Tue, 20 Sep 2022   Prob (F-statistic):           1.51e-25
Time:                        21:52:26   Log-Likelihood:                -1484.1
No. Observations:                 163   AIC:                             2974.
Df Residuals:                     160   BIC:                             2984.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                -3177.5539 

C:\Users\Ankit\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning:

In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only



Finally we find two indenpedent variables are statistically significant but still they are able to explain 51% of variation in revenue.

In [115]:
X=dataframe[variables]
y=dataframe['revenue']
X_train, X_test, y_train , y_test = train_test_split(X, y, shuffle = False, test_size=0.2)

model = LinearRegression()
model.fit(X_train, y_train)

print(model.score(X_train, y_train), model.score(X_test, y_test))

0.5069285376363957 0.5050868157140338


We have split the data in ratio 80:20 and then build a linear regression model. then we calculate R-squared values of both trainand test set. We have not shuffled the data in this regression model.

**Test and Train set, both yield similar R-squared values**

In [124]:
X=dataframe[variables]
y=dataframe['revenue']
X_train, X_test, y_train , y_test = train_test_split(X, y, shuffle = False, test_size=0.5)

model = LinearRegression()
model.fit(X_train, y_train)

print(model.score(X_train, y_train), model.score(X_test, y_test))

0.4520154379838137 0.5082221582700106


In [125]:
model.coef_

array([ 1.13336763, 50.24288596])

In [126]:
X.columns

Index(['google_organico', 'fb_prospecting_spent'], dtype='object')

In [127]:
dataframe['prediction']=model.predict(X)

fig = px.line(dataframe, x='week', y=dataframe.columns)
fig.show()

Here we can see that apart from two instances, The prediction of the model is almost close the actual value.

Cost per Revenue (CPR)

In [130]:
google_organico = 1/1.13336763
google_organico

0.8823262404273889

In [131]:
fb_prospecting_spent = 1/50.24288596
fb_prospecting_spent

0.01990331528320512

**Adstock affect and Diminishing return**

transfer. = (adstock(t-1)*beta+spent(t))^alpha

Dim Ret. = Adstock(t)^Alpha

Where Value of Alpha and beta is to be estimated

For start

Let's say:

alpha = 1
Beta = .1

In [170]:
dataframe = pd.read_csv('dataset_real.csv')

In [171]:
beta=.1
alpha=1
variables_list=['fb_prospecting_spent','google_organico']
for j in range (2):
    for i in range (162):
        if i>0:
            dataframe[variables_list[j]][i]=dataframe[variables_list[j]][i]*beta+dataframe[variables_list[j]][i]

C:\Users\Ankit\AppData\Local\Temp/ipykernel_28744/1191028802.py:7: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [172]:
dataframe[['fb_prospecting_spent','google_organico']].head()

fb_prospecting_spent  google_organico
0                31.480             2603
1                32.571             6260
2                33.099             6177
3                33.088             6243
4                34.133             5937

In [173]:
X=dataframe[variables_list]
y=dataframe['revenue']
X = sm.add_constant(X)

result = sm.OLS(y,X).fit()

print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                revenue   R-squared:                       0.510
Model:                            OLS   Adj. R-squared:                  0.504
Method:                 Least Squares   F-statistic:                     83.27
Date:                Tue, 20 Sep 2022   Prob (F-statistic):           1.64e-25
Time:                        23:41:24   Log-Likelihood:                -1484.2
No. Observations:                 163   AIC:                             2974.
Df Residuals:                     160   BIC:                             2984.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                -3154.5325 

C:\Users\Ankit\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning:

In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only



In [186]:
dataframe = pd.read_csv('dataset_real.csv')
beta=.1
alpha=1
for i in range (162):
    if i>0:
        dataframe['google_organico'][i]=dataframe['google_organico'][i]*beta+dataframe['google_organico'][i]
beta=.5
alpha=.5
for i in range (162):
    if i>0:
        dataframe['fb_prospecting_spent'][i]=dataframe['fb_prospecting_spent'][i]*beta+dataframe['fb_prospecting_spent'][i]

C:\Users\Ankit\AppData\Local\Temp/ipykernel_28744/1271334510.py:6: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

C:\Users\Ankit\AppData\Local\Temp/ipykernel_28744/1271334510.py:11: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [187]:
X=dataframe[variables_list]
y=dataframe['revenue']
X = sm.add_constant(X)

result = sm.OLS(y,X).fit()

print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                revenue   R-squared:                       0.509
Model:                            OLS   Adj. R-squared:                  0.503
Method:                 Least Squares   F-statistic:                     83.02
Date:                Tue, 20 Sep 2022   Prob (F-statistic):           1.86e-25
Time:                        23:52:58   Log-Likelihood:                -1484.3
No. Observations:                 163   AIC:                             2975.
Df Residuals:                     160   BIC:                             2984.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                -3112.6724 

C:\Users\Ankit\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning:

In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only



So when we change the alpha (carry-over) and beta (ad-stock) affect, R-squared value of model donot change much. 

# Overall we conclude that there are two main contributing marketing channels: fb_propecting_spent and google_organico